### PR TITLE
[Fiber] Mount/unmount warnings and invariants

### PR DESCRIPTION
**Pull request by acdlite**
Fixes some of the tests in `ReactMount-test.js`, except the ones related to server-side rendering.

### CODE REVIEW

File: scripts/fiber/tests-failing.txt [hunk 1]
```diff
@@ -42,6 +40,17 @@ src/renderers/shared/__tests__/ReactPerf-test.js
 * should work when measurement starts during reconciliation
 
 src/renderers/shared/hooks/__tests__/ReactComponentTreeHook-test.js
+* ignores null children
```
Comment (gaearon) [inline scripts/fiber/tests-failing.txt]: Why have they moved to failed?

Comment (acdlite) [inline scripts/fiber/tests-failing.txt]: Just noticed that, not sure why

**Comment by acdlite** [comment]
@gaearon Ok fixed it

File: src/renderers/dom/fiber/ReactDOMFiber.js [hunk 2]
```diff
@@ -368,7 +380,43 @@ var ReactDOM = {
...
+          element != null && element.props !== undefined ?
+            ' This may be caused by unintentionally loading two independent ' +
+            'copies of React.' :
+            ''
+        )
+      );
+    }
+
+    if (__DEV__) {
+      const isRootRenderedBySomeReact = Boolean(container._reactRootContainer);
```
Comment (spicyj) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: Can you verify the logic here? This means if you call render on a container with existing React children created by another React, we don't warn? I guess that is consistent with before.

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js [hunk 3]
```diff
@@ -673,6 +673,7 @@ describe('ReactUpdates', () => {
   it('should flush updates in the correct order across roots', () => {
     var instances = [];
     var updates = [];
+    var container = document.createElement('div');
```
Comment (spicyj) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: Why the change here?

Comment (sebmarkbage) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: Hm. This looks wrong. This changes the behavior we're testing. Why doesn't it work?

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js [hunk 4]
```diff
@@ -681,20 +682,21 @@ describe('ReactUpdates', () => {
       }
 
       componentDidMount() {
+        this.container = document.createElement('div');
```
Comment (sebmarkbage) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: Specifically we want to test that we can render into the node we already rendered.

Comment (acdlite) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: It does work the old way but it warns. Stack doesn't warn but it seems like it should?

Why doesn't Stack warn?

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js (hunk 3 above)
Comment (acdlite) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: https://github.com/facebook/react/pull/8919#discussion_r99279560

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js (hunk 4 above)
Comment (spicyj) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: It's allowed to render into an empty element. This test intentionally tests that so your change messes that up.

File: src/renderers/dom/fiber/ReactDOMFiber.js (hunk 2 above)
Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: Yeah we currently only warn about different copies of React on unmount

**Comment by acdlite** [comment]
@spicyj Fixed

File: src/renderers/dom/fiber/ReactDOMFiber.js (hunk 2 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: Do these strings get compressed? Would it be better if we split them into separate invariants?

File: src/renderers/dom/fiber/ReactDOMFiber.js [hunk 5]
```diff
@@ -368,7 +380,44 @@ var ReactDOM = {
...
+            'React.createElement(\'div\') or <div />.' :
+          typeof element === 'function' ?
+            ' Instead of passing a class like Foo, pass ' +
+            'React.createElement(Foo) or <Foo />.' :
+          // Check if it quacks like an element
+          element != null && element.props !== undefined ?
+            ' This may be caused by unintentionally loading two independent ' +
+            'copies of React.' :
+            ''
+        )
```
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: Do these strings get compressed? Would it be better if we split them into separate invariants?

Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: It would be nice if error code transform was throwing on anything other than static messages.

Comment (spicyj) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: @gaearon: It does; this is a %s param so you can put whatever you want there.

Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: Yeah, I copy and pasted this from Stack. I'll change it in both places to use multiple invariants instead.
